app/lib/search/backend.dart

```diff
@@ -138,6 +147,71 @@ class SearchBackend {
   }
 }
 
+class PopularityStorage {
+  final Storage storage;
+  final Bucket bucket;
+
+  final Map<String, double> _values = {};
+
+  PopularityStorage(this.storage, this.bucket);
+
+  double operator [](String package) => _values[package];
+
+  Future init() async {
+    await fetch();
+    new Timer.periodic(const Duration(days: 1), (_) {
+      fetch();
+    });
+  }
+
+  Future fetch() async {
+    _logger.info('Updating popularity storage...');
+    final List<BucketEntry> list = await bucket.list().toList();
```

Comment:
After time this will become a quite slow operation. Instead of listing the entire bucket contents, which can not only do several RPCs (under the hood), we should only try to fetch a `popularity-latest.json`.

The uploader should be able to upload as `popularity-YYYY-MM-DD.json` (so we have the historic values) and then do a compare-and-swap way to copy it over `popularity-latest.json`. There will be no races and everything is atomic: Either the new file is there entirely or the old file is there, this is guaranteed by GCS.


(Same is true for snapshots of search index: If we cache it on GCS I have probably added the same comment there)

Reply:
Fair point. I'd like to keep the current logic until @kevmoo confirms that the upload process will do the swap. In a separate PR I'll update the snapshot code to use the same thing.

Btw. I think it may be worth to keep the full list fetch as a fallback if the file is not there for some reason. WDYT?

Reply:
To ensure this gets addressed I've filed https://github.com/dart-lang/pub-dartlang-dart/issues/468